[3단계 - 리팩터링] 플린트(이주환) 미션 제출합니다.

jhan0121 opened the pull request:
안녕하세요 돔푸!
이번 코드 리뷰도 잘부탁드립니다! 아래는 주요 변경 사항입니다. 이 외에도 눈에 띄는 부분이 있으시다면 언제든 코멘트 달아주세요!!

# 주요 변경 사항
## HttpCookie 생성 메서드의 추상화 레벨 문제

step2에서 돔푸가 "추상화 레벨이 깨져보인다"는 말씀에 추가적인 리팩터링 작업을 수행했습니다. 이번에는 추상화 레벨은 어느정도 지켜진 느낌이 들었지만 `parseCookies` 메서드 부분에 if문이 좀 많아 보여서 추가적으로 개선할 만한 요소가 있거나, 돔푸가 생각했던 의도는 어떤 방향으로 진행하는 것이었는지 알면 더 좋은 구조로 개선할 수 있을 것 같습니다!

```
from(httpHeader 객체)
│
├──  Cookie 헤더 존재 시
│   │
│   └── parseCookies(쿠키 문자열) 호출
│       │
│       ├── 1. ";" 기준으로 문자열 분리
│       │
│       └── 2. 각 쌍("key=value")을 순회하며
│           │
│           └── "=" 기준으로 키/값 분리 후 검증 통과 시, Map에 저장
│
│   └── 최종: 파싱된 Map으로 HttpCookie 객체 생성
│
└──  Cookie 헤더 미존재 시
    │
    └── 최종: 빈 Map으로 HttpCookie 객체 생성
```

## HttpRequest와 HttpRequestBody 사이의 책임 정리

HttpRequest 생성 로직에서 HttpRequestBody만 bufferedReader를 파라미터로 받는 형식을 개선했습니다. "buffferedReader로 데이터를 읽는 책임"을 HttpRequest에 위임하고 HttpRequestBody는 "값만 세팅해서 갸지는 책임"만 남겨서 일관성을 챙겼습니다! 이 외에도 고려할만한 요소가 있다면 언제든지 말씀해주세요!

## Controller 인터페이스를 통한 if문 지옥 개선하기

이번 step에서 드디어 controller 메서드 분기 처리를 개선했습니다.

- `controller`라는 인터페이스와 `AbstractController`리는 추상 클래스를 구현해서 "정적 파일 처리", "로그인", "회원 가입" 로직을 각각 처리하는 Controller를 구현했습니다.

http 요청이 오면, 어떤 컨트롤러를 사용해야 하는지 판단하는 로직은 `ControllerProvider`를 활용해서 path 일치 여부로 대상 controller를 찾도록 구현했습니다.

새로운 controller 등록이 필요하면 `ControllerProvider` 에 추가해주기만 하면 나머지 코드는 수정할 필요가 없어진다는 장점이 있었습니다.

Commit:
fix(study): Spring Boot 의존성 관리로 JUnit 정렬 및 테스트 오류 해결

- junit-jupiter-api/engine 직접 버전 제거
- spring-boot-starter-test로 테스트 의존성 일원화
- AssertJ/Mockito 버전 지정 제거해 Boot 관리로 위임
- JUnit Jupiter/Platform 불일치로 인한 NoSuchMethodError 해소

jhan0121 committed:
Merge branch 'woowacourse:jhan0121' into jhan0121

jhan0121 committed:
fix: index.html 상단 바 로그인 블럭 수정

jhan0121 committed:
feat: Controller, AbstractController 추가

jhan0121 committed:
feat: LoginController 추가

jhan0121 committed:
feat: RegisterController 추가

jhan0121 committed:
feat: StaticResourceController 추가

jhan0121 committed:
feat: InternalServerErrorException 추가

jhan0121 committed:
refactor: HttpStatus 500 추가

jhan0121 committed:
refactor: UnauthorizedException 삭제

jhan0121 committed:
feat: ControllerProvider 추가

jhan0121 committed:
feat: http 응답 로직 개선

- ControllerProvider에서 path 기반 controller 탐색 기능 추가
- controller#service 실행 기능 추가

jhan0121 committed:
refactor: HttpController 삭제

jhan0121 committed:
refactor: LoginService 추가

- 유저 로그인 기능 LoginService로 이동

jhan0121 committed:
refactor: RegisterService 추가

- 유저 가입 기능 RegisterService로 이동

jhan0121 committed:
style: 코드 포맷 정리

jhan0121 committed:
refactor: 불필요한 메서드 삭제

jhan0121 committed:
fix: 로그인 실패 처리 try-catch 적용

jhan0121 committed:
Merge branch 'woowacourse:jhan0121' into jhan0121

jhan0121 committed:
refactor: HttpCookie 불필요한 메서드 삭제

jhan0121 committed:
refactor: HttpCookie 생성 로직 수정

jhan0121 committed:
refactor: HttpRequest 생성 로직 수정

- HttpRequest에 bufferedReader 라인 읽기 책임 위임
- HttpRequestBody에 bufferedReader 라인 읽기 책임 제거

jhan0121 committed:
refactor: 대상 controller 탐색 실패 시, 404로 응답하도록 변경

jhan0121 committed:
refactor: 기본 응답 status 설정 로직 추가

jhan0121 committed:
feat: NotFoundException 추가

dompoo commented on tomcat/src/main/java/org/apache/coyote/http11/controller/Controller.java:
```diff
@@ -0,0 +1,11 @@
+
+    boolean isProvide(String path);
+
+    void service(HttpRequest request, HttpResponse response);
```
request와 response를 모두 파라미터로 넘기는 방식은 사실 흔하지 않은데요,
보통 response를 return 값으로 사용하기 때문입니다.
이렇게 구성하게된 이유가 궁금해요~!!

jhan0121 replied:
controller에서 http 관련 요소(session, cookie 등)을 직접 만들어서 사용하는 방식보다 이미 제공된 Response에 "cookie는 ~~로 보내주시고 session에 ~~ 넣어주세요"로 요청을 하는 것이 더 안전하지 않을까?하는 생각이 들어 만들어진 response를 이용하도록 구성해보았습니다!

dompoo replied:
안전하다는건 어떤건가요? tomcat이 터지지 않는다는 건가요? 아니면 보안 유출?

jhan0121 replied:
response를 만드는 사용자 입장에서 response 생성에 대한 안정성을 높일 수 있다는 느낌이었습니다!

예를 들어 "session key-value를 추가하는 상황"이라고 가정했을 때, Response객체를 직접 생성해서 리턴하는 방식이라면

생성자에 넣어주어야 하는 정보로

- header 정보
- cookie 정보 (cookie를 사용한다면 header에 set-cookie 헤더 추가)
- session 정보 (session을 사용한다면 cookie에 jsession 등록 필요)

를 수행해서 Resonse를 생성하고 리턴하는 방식이 될 것 같은데, 이 방식에서 response 생성을 위한 요소들 간의 조합 경우의 수(body가 있는지/cookie가 있는지/session이 있는지 여부 등)가 많아져서 사용자가 response 생성하는 과정이 복잡해지지 않을까 하는 생각이 있었습니다.

dompoo replied:
사용자가 req res를 생성하는 과정이 복잡해지니, 그 과정에서 휴먼에러가 발생하기 쉬워진다는 뜻이군요.

그래서 어느정도 이미 구성된 res를 넘겨서 이를 해결하시려고 한 시도라고 이해했어요.

하지만 그것만으로는 기존 코드와 다른 형태로 req res를 모두 넘겨주는 것의 이유가 되지는 못하는게, 기존 코드처럼 작성해도 꽤나 안전하게 작성할 수 있거든요.

`Response.body("index.html");` -> 200OK, indexhtml이 바디고 컨텐트 타입/길이가 적절히 설정

`Response.setCookie(...)` -> 세션id쿠키 응답

이런 형태로 작성해도 될 것 같거든요.

---

사실 실제 서블릿도 플린트의 코드와 형태가 비슷해요. 그렇게 구성된 이유를 찾아서 공부해보고 그 이유가 플린트의 상황에 적용하기 알맞은지까지 알려주세요.

dompoo replied:
제가 이렇게까지 여기에 고민하는 이유는, 이런식으로 구성되면 다형성을 전혀 활용하지 못할 것 같기 때문이에요.

`return SimpleResponse.body("index.html")`
`return MergedResponse.of(SimpleResponse.body(...), SessionCookieResponse.jsession())`
`return new MyErrorResponse(INVALID_EMAIL);`
...

정말 수많은 방법으로 res를 생성하고 조합하는 것을 상상하면 확장 포인트가 많아요. 이런것들을 포기한 이유가 뭘까요?

dompoo commented on tomcat/src/main/java/org/apache/coyote/http11/controller/Controller.java:
```diff
@@ -0,0 +1,11 @@
+
+public interface Controller {
+
+    boolean isProvide(String path);
```
path에 따라서 지원하는지 결정되는거라면, 사실 해당 톰캣을 사용하는 애플리케이션 개발자 입장에서는
이렇게 메서드 시그니처를 구성하는 것보다는 `String providableUrl()` 같은 형태가 더 사용하기 좋을 것 같아요.
이렇게 구성하시게된 이유가 궁금합니다.

jhan0121 replied:
처음에는 "이 경로를 지원하는 컨트롤러인가?"라는 의미를 표현하는 방법을 생각해보면서 파라미터에 Path라고 명시되어 있으니까 `isProvide`면 충분하지 않을까? 하는 생각에 `isProvide`로 작성했었습니다.

여기서 돔푸의 의견을 듣고 보니, 현재 시그니처만 딱 보았을 때, "무엇을 지원한다고 작성해야 하지?"라고 생각할 가능성이 있을 것 같네요! 물론 파라미터명을 `urlPath`로 변경하는 것으로 충분할 수 있지만,  시그니처명으로 더 명확하게 확인할 수 있게 해주는 것이 제가 사용자 입장이라면 더 편할 것 같다고 생각이 들어서 아래 형태로 수정했습니다~

```java
    boolean isProvidableUrl(String urlPath);
```

https://github.com/woowacourse/java-http/pull/959/commits/3f9b10691eee0ccd597f8a10e75dc58528fa8484

dompoo replied:
앗, 메서드명을 이야기하는게 아니에요.

컨트롤러 입장에서
url을 받아서 해당 url이 내가 처리할 수 있는지 판단하는게 현재 로직인데,
그러지말고 그냥 컨트롤러가 자기가 처리할 수 있는 url의 정규식 같은 것을 응답하는게 애플리케이션 개발자 입장에서 더 구현하기 쉽지 않을까요?

jhan0121 replied:
아하, 확실히 ControllerProvider가 controller에 정의된 경로를 보고 판단하는 형태가 더 객체지향적인 구조네요!

여기서, StaticResourceController의 경우에는 사용자가 경로를 정의해줄 필요 없이, tomcat에서 기본 제공하는 controller라고 보는 것이 더 맞다고 생각이 들어서 tomcat의 controller 패키지로 이동시켰습니다.

https://github.com/woowacourse/java-http/pull/959/commits/653511625906ce7750bc07ef4fb367d6612158c9

dompoo commented on tomcat/src/main/java/org/apache/coyote/http11/controller/AbstractController.java:
```diff
@@ -0,0 +1,22 @@
+import org.apache.coyote.http11.http.request.HttpRequest;
+import org.apache.coyote.http11.http.response.HttpResponse;
+
+public abstract class AbstractController implements Controller {
+
+    @Override
+    public void service(HttpRequest request, HttpResponse response) {
+        if (request.getMethod().equals(HttpMethod.GET)) {
+            doGet(request, response);
+        }
+        if (request.getMethod().equals(HttpMethod.POST)) {
+            doPost(request, response);
+        }
+    }
+
+    protected abstract void doGet(HttpRequest request, HttpResponse response);
+
+    protected abstract void doPost(HttpRequest request, HttpResponse response);
```
이런 컨셉 아주 좋아합니다 ㅋㅋ
`AbstractController` 덕분에 개발자는 한층 더 쉽게 개발할 수 있겠네요.
실제로 다른 라이브러리를 만질 때에도, `AuthenticationManager` 같은건 구현하기 막막한데, `UsernamePasswordAuthenticationManager` 를 적절히 상속해서 손쉽게 구현해낸 기억이 있거든요.

jhan0121 replied:
중간에 사용하기 편한 구현체가 하나 있으면 더 편할 것 같아서 추가해보았습니다!

dompoo commented on tomcat/src/main/java/org/apache/coyote/http11/controller/ControllerProvider.java:
```diff
@@ -0,0 +1,26 @@
+            new RegisterController(new RegisterService()),
+            new LoginController(new LoginService()),
+            new StaticResourceController()
+    );
```
톰캣이 애플리케이션 개발자가 개발한 컨트롤러에 의존하는 형태인데, 이거 괜찮을까요?

jhan0121 replied:
해당 부분에도 문제가 있습니다!! 이를 해결하기 위해 사용자의 controller를 등록하는 설정 로직을 추가해서 사용자가 직접 등록하도록 수정했습니다.

https://github.com/woowacourse/java-http/pull/959/commits/c0075dbff8da9f67cbf73e065727901f82f56307

dompoo replied:
넵 좋습니다~!!
한가지 더 첨언하자면,(수정요청X) `ControllerProvider`라는 이름은 온전히 톰캣 개발자의 관점에서 작성된 것 같은데요, 애플리케이션 개발자 입장에서 `ControllerProvider`에 컨트롤러를 끼워넣는 것은 조금 이상하게 느껴질 수도 있어요.
(컨트롤러 제공자..? 뭘 제공하지? 내가 개발하는건데..? 여기에 컨트롤러를 넣으라고?)

jhan0121 replied:
맞습니다!! 해당 문제는 아직 완전한 해결책이 아니라고 생각이 들어서 아쉬운 요소 중 하나였습니다. spring mvc와 비슷하게 사용자는 컨트롤러라는 표시를 추가하는 것 만으로도 이를 감지해서 자동으로 등록할 수 있는 기능을 만든다면 문제를 해결할 수 있을 것 같네요!!

dompoo replied:
좋아요 mvc 들어가서 더 해봅시다

dompoo commented on tomcat/src/main/java/com/techcourse/controller/StaticResourceController.java:
```diff
@@ -0,0 +1,49 @@
+
+        final String resourcePath = path.substring(1);
+        response.setOk(resourcePath);
+    }
```
코드가 조금 복잡한데요, 제가 생각하는 주원인은 "/"에 대한 응답과 정적 리소스에 대한 응답이 혼재되어 있기 때문인 것 같아요.

애초에 `isProvide()` 라는 메서드를 작성해주신 것은, `doGet()` 에서 if문을 최소화하기 위함이라고 이해했거든요. `isProvide()` 를 제대로 작성하지 않는다면, 사실상 `isProvide()` 에서는 항상 true를 응답하고, `doGet()`에서 if문을 길~게 작성해주면 되니까요.

컨트롤러의 책임을 적절히 나누어보면 좋을 것 같습니다!

jhan0121 replied:
확실히 "/"를 처리하는 로직과 정적 리소스를 처리하는 로직을 분리하는 것이 더 의도에 맞는 구조일 것 같네요! "/"에 대한 처리를 위한 별도의 RootController 를 구성하였습니다!

https://github.com/woowacourse/java-http/pull/959/commits/1ba648eef36bc4ec352d7fd4d29a679946274086

dompoo replied:
간단하고, 코드 중복 같지만 훨씬 명시적이라 좋네요!

dompoo commented on tomcat/src/main/java/com/techcourse/controller/StaticResourceController.java:
```diff
@@ -0,0 +1,49 @@
+    protected void doGet(final HttpRequest request, final HttpResponse response) {
+        final String path = request.getPath();
+
+        if (path.contains("..")) {
```
이 부분은 순수하게 이해가 안되는데, ..이 포함되는 경우가 존재하나요...?

jhan0121 replied:
클라이언트 측에서 악의적으로 `/static/../../../{숨겨둔 파일}`과 같은 형태로 처리된다면 static 디렉터리 바깥쪽을 접근할 가능성이 있을 것 같다는 생각이 들어서 추가해보았습니다!

dompoo replied:
오호 이해했습니다. 그런게 아래 코드에서 실제로 가능한가요?
테스트 해보고, 결과 알려주시면 감사하겠습니다.

```java
public static HttpResponseBody withStaticResourceName(String resourceName) {
    final URL url = HttpResponseBody.class.getClassLoader().getResource("static/" + resourceName);

    if (url == null) {
        return withString(resourceName);

    }

    try {
        final Path path = Paths.get(url.toURI());
        return new HttpResponseBody(Files.readAllBytes(path),
                ContentTypeValue.findByTarget(Files.probeContentType(path)));
    } catch (IOException | URISyntaxException e) {
        throw new IllegalArgumentException("response 구성 중 문제가 발생하였습니다", e);
    }
}
```

jhan0121 replied:
공격 방지 코드를 주석 처리하고 curl로 직접 실험해본 결과, static 바깥의 파일을 다운로드에 성공하는 것을 확인할 수 있었습니다!
(크롬이나 postman은 `..` 라인을 검증해서 삭제하는 처리를 수행해서 아래의 curl 명령어로 테스트해보았습니다.)


### 공격 방지 코드 주석 처리
```java
public class StaticResourceController extends AbstractController {
...
    @Override
    protected void doGet(final HttpRequest request, final HttpResponse response) {
        final String path = request.getPath();

        // 경로 조작 공격 방지
//        if (path.contains("..")) {
//            log.warn("directory scan attack detected! : {}", path);
//            response.setNotFound();
//            return;
//        }

        final String resourcePath = path.substring(1);
        response.setOk(resourcePath);
    }
...
```

### intelij로 실행 후, curl 커맨드 실행
```cmd
curl --path-as-is "http://localhost:8080/../test_secure.svg" -o stolen.svg
```

### 실행 결과
<img width="2559" height="1379" alt="image" src="https://github.com/user-attachments/assets/bd209f02-a50e-4865-beee-728370443ae3" />

https://github.com/woowacourse/java-http/pull/959/commits/475ecefdec12a3af4d6088287c5483e6e1abcb0c

dompoo replied:
홀리몰리~ 잘 테스트해주셨네요. (버그인지 모르겠지만 모바일에서 이미지가 안보이네요 ㅋㅋ)

이거 분명 불편한 포인트거든요? 악성 URI 객체를 통해서 데이터를 뽑아 먹을 수 있다. -> 안전한 URI는 없을까? 까지 공부해봅시다!

dompoo commented on tomcat/src/main/java/com/techcourse/controller/RegisterController.java:
```diff
@@ -0,0 +1,39 @@
+    @Override
+    protected void doGet(final HttpRequest request, final HttpResponse response) {
+        response.setOk("register.html");
+    }
```
해당 컨트롤러에서는 `GET /register` 에 대한 정적 리소스 응답과, `POST /register` 에 대한 동적 로직이 같이 구성되어 있네요.
이미 `StaticResourceController` 가 잘 작성되어 있는데, 이걸 활용하지 않은 이유가 궁금해요!

jhan0121 replied:
`index.html`과 같은 형태의 경우에는 `StaticResourceController`를 통하는 것이 맞지만, `/register`나 `/login`과 같은 케이스는  "/register 경로로 요청이 오면 register.html을 전송해줘"와 같이 특정 경로와 매핑을 하는 경우라고 보아서 별도로 doGet을 구성했습니다.

dompoo replied:
이런식이면 정적 리소스가 어떻게 응답되는지 추적하기가 너무 어려울 것 같은데요...?
애플리케이션 개발자가 컨트롤러 10개쯤 구현했을 때, "우리 정적 리소스 뭐 응답하고 있어?" 라는 질문에 답할 수 있을까요?

dompoo replied:
톰캣과 우아코스 디렉토리의 코드는 전혀 다른 관점으로 개발되어야 한다고 생각해요.
톰캣 디렉토리의 코드는 최대한 안전하고, 사용하기 쉽게 개발되어야 하고
우아코스 디렉토리의 코드는 최대한 명시적이고, 읽기 쉽게 개발되어야 해요.
이런 관점으로 컨트롤러의 구조를 좀 더 고민해보면 좋을 것 같습니다.

jhan0121 replied:
확실히 지금 구조로는 일일이 컨트롤러 클래스들을 각각 접근해서 정적 리소스 응답 여부를 확인해야 하는 문제가 있을 것 같네요...!

하나의 Controller 클래스에 여러 응답 메서드를 둘 수 있다면 좀 더 편하게 확인할 수 있을 것 같은데 spring mvc와 형태가 비슷한 방향으로 구조를 개선하는 방식이 있을 것 같습니다!

또 하나의 방법으로는 돔푸의 말씀대로 static 파일에 응답하고자 하는 리소스만 저장하고 `StaticResourceController`에 모든 책임을 위임하는 방법도 있지만 `localhost:8080/login` 등의 특정 경로 요청을 어떻게 `login.html` 등의 정적 파일로 매핑해야 할지 고민입니다.

혹시 이 부분에 대해 생각하신 좋은 방법이 있다면 조언 부탁드립니다!

dompoo replied:
`interface Controller` : 최상단 인터페이스
`class SimpleStaticResourcesController` : 컨트롤러 구현한 구체클래스, 여러 정적 리소스를 특별 로직 없이 응답
`abstract StaticResourceController` : 컨트롤러 구현한 추상클래스, 정적리소스 서빙에 로직이 들어갈 때 구현하여 사용


정도로 구분해도 될 것 같고, 방법은 정말 많을 것 같아요.

dompoo replied:
이 부분은 고민만 하되 따로 수정하지 마시고, mvc 들어가서 더 고민해보는 것도 좋아보여요!

dompoo commented on tomcat/src/main/java/org/apache/coyote/http11/http/response/HttpResponse.java:
```diff
@@ -75,12 +55,40 @@ private List<String> getResponseLines() {
+        this.header.addHeader(HttpHeaderKey.LOCATION.getValue(), targetPath);
+    }
+
+    public void setUnauthorized() {
+        this.responseLine = HttpStatusLine.of(HttpVersion.HTTP_1_1, HttpStatus.UNAUTHORIZED);
+        this.responseBody = HttpResponseBody.withStaticResourceName("401.html");
+        this.header = HttpHeader.createByResponseBody(this.responseBody, "401.html");
+    }
+
+    public void setNotFound() {
+        this.responseLine = HttpStatusLine.of(HttpVersion.HTTP_1_1, HttpStatus.NOT_FOUND);
+        this.responseBody = HttpResponseBody.withStaticResourceName("404.html");
+        this.header = HttpHeader.createByResponseBody(this.responseBody, "404.html");
+    }
+
+    public void setInternalServerError() {
+        this.responseLine = HttpStatusLine.of(HttpVersion.HTTP_1_1, HttpStatus.INTERNAL_SERVER_ERROR);
+        this.responseBody = HttpResponseBody.withStaticResourceName("500.html");
+        this.header = HttpHeader.createByResponseBody(this.responseBody, "500.html");
     }
```
401, 404, 500 html 은 tomcat의 기본 응답인거죠?
저는 woowacourse쪽 리소스라고 생각했는데, 관점이 달라서 신기하네요~!!

뭔가 이렇게 되니, html을 모든 Status에 대해서 만들지 말고, 동적으로 내용만 살짝 바꿔줄 수도 있겠다는 생각이 드네요! (당연히 제안은 아닙니다 ㅎㅎ)

jhan0121 replied:
해당 html파일을 사용자가 만들어 놓으면 자동으로 응답되게 하면 좋을 것 같다는 생각이 들어서 적용해보았습니다ㅎㅎ

dompoo replied:
오잉 사용자가 만드는건가요? ㅋㅋ 그러면 좀 더 명시적으로 문서를 남겨주는게 좋을 것 같아요.
저는 401 400 같은 것들은 톰캣수준에서 기본값으로 반환해주는거고, 사용자에게 커스터마이징 포인트를 주는줄 알았습니다.

jhan0121 replied:
spring에서 static/error 디렉터리에 `404.html` 등을 추가해두면 알아서 리턴해주는 기능을 따라해 본 느낌이었어요!! 확실히 처음 사용하는 사용자라면 이런 기능이 있는지 잘 모를 수 있어서 별도의 문서화는 필요할 것 같습니다!

dompoo commented on tomcat/src/main/java/org/apache/coyote/http11/http/common/HttpCookie.java:
```diff
@@ -1,62 +1,56 @@
-        httpCookie.addCookie(cookieName, cookieValue);
+        return values;
     }
 
```
stream으로 깔끔하게 뽑아낼 수 있을 것 같아요!
어떻게 하는게 좋을까용?

jhan0121 replied:
stream이 더 가독성이 높아보이네요!! stream을 사용하는 로직으로 수정하였습니다!

https://github.com/woowacourse/java-http/pull/959/commits/b0301c0f2ade885abf1126dcbbee11ce0064b034

dompoo reviewed (changes requested):
안뇽하세요 플린트 이번 스텝도 고생 많으셨어요!!

전체적으로 리팩터링 잘해주셨는데, 특히 Controller-AbstractController 쪽에서 큰 변화가 있더라고요 ㅎㅎ 아주 좋습니다.

뭔가 시간이 지나면서 점점 플린트의 쪼가 익숙해지다보니까 리뷰할 내용도 점점 적어지는 것 같아요.ㅋㅋ

의견 언제든지 편하게 주시고, 제가 드리는 모든 피드백은 이렇게 고치라! 라는게 아니라 얘기해보자~ 라는거니까, 코멘트 남겨주세요!

jhan0121 committed:
refactor: Controller 지원 path 설정 메서드 시그니처 수정

jhan0121 committed:
refactor: 사용자 controller 등록 로직 추가

jhan0121 committed:
refactor: 로그인/회원 가입 컨트롤러 로직 수정

- session 무효화 기능 HttpRequest로 위임
- body element 검색 기능 HttpRequest로 위임

jhan0121 committed:
refactor: 루트 경로 처리 컨트롤러 분리

jhan0121 committed:
refator: HttpCookie 파싱 로직 stream 적용

jhan0121 committed:
fix: 쿠키 파싱 로직 오류 수정

dompoo reviewed (changes requested):
기존 코멘트에 답글만 좀 달아놨습니다! 더 얘기할 필요가 있어보여요.

jhan0121 committed:
test: 테스크 클래스에 사용자 Controller 등록 로직 추가

jhan0121 committed:
refactor: HttpResponseBody#withStaticResourceName 로직 수정

jhan0121 committed:
test: 경로 조작 공격 테스트용 데이터 추가 및 공격 감지 로그 추가

jhan0121 committed:
refactor: 대상 Controller 탐색 로직 개선

- Controller#providableUrl 추가
- ControllerProvider#findByPath 수정
- DefaultControllerConfig 추가

dompoo commented:
안녕하세요 플린트- step4가 짧긴한데, 그래도 시간이 별로 없어서 이만 머지할게요. 기존 코멘트들에 제 의견을 더 첨부했으니 확인하고, 궁금한 것이 있다면 디엠으로 부탁드려요!
